### PR TITLE
Introduce @CombineTestTemplates support

### PR DESCRIPTION
**Pull request by danielhodder**
This adds support for combining multiple TestTemplate providers
together, in a product style, so that an implementer can use the
benefits of multiple extensions together.

Issue: #1224

## Overview

This is a fix for #1224. I had some free time so I thought I would take a look at how hard it would be. This will almost certainly not work for everyone so I have made it opt-in with a new annotation, @CombineTestTemplates. This will allow those who need this feature to use it, and everything will be normal for others.

Using the Maven test sample as a quick test we get the following test output (happy to hear suggestions for better test names): 
image

I have not yet had a chance to write test cases but I'm gonna have a crack at that later in the week if this seems interesting.

---

I hereby agree to the terms of the [JUnit Contributor License Agreement](https://github.com/junit-team/junit5/blob/002a0052926ddee57cf90580fa49bc37e5a72427/CONTRIBUTING.md#junit-contributor-license-agreement).

---

### Definition of Done

- [x] There are no TODOs left in the code
- [x] Method [preconditions](https://junit.org/junit5/docs/snapshot/api/org.junit.platform.commons/org/junit/platform/commons/util/Preconditions.html) are checked and documented in the method's Javadoc
- [x] [Coding conventions](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#coding-conventions) (e.g. for logging) have been followed
- [x] Change is covered by [automated tests](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#tests) including corner cases, errors, and exception handling
- [x] Public API has [Javadoc](https://github.com/junit-team/junit5/blob/HEAD/CONTRIBUTING.md#javadoc) and [`@API` annotations](https://apiguardian-team.github.io/apiguardian/docs/current/api/org/apiguardian/api/API.html)
- [x] Change is documented in the [User Guide](https://junit.org/junit5/docs/snapshot/user-guide/) and [Release Notes](https://junit.org/junit5/docs/snapshot/user-guide/#release-notes)
- [x] All [continuous integration builds](https://github.com/junit-team/junit5#continuous-integration-builds) pass

### CODE REVIEW

**Comment by sbrannen** [comment]
Tentatively slated for 5.8 M1 solely for the purpose of _team discussion_

**Comment by SpencerMelo** [comment]
Any idea on when we are going to be able to use it?

**Comment by parnmatt** [comment]
I too am quite interested in if/when this would be accepted. 
The assignee and reviewer lists are still empty; so could be a while.

**Comment by mikalaikarol** [comment]
Looking forward for it as well

**Comment by mikalaikarol** [comment]
sad that there was no continuation on this PR

**Comment by parnmatt** [comment]
> sad that there was no continuation on this PR

Indeed, I was hoping to use this feature at some point soon.

**Comment by SpencerMelo** [comment]
It's really sad, the code was there. I don't understand why it will not move forward, for folks writing large scale tests, with a lot of data, reporting, different stuff going on, this would be useful, teams that are using junit5 at it full extents often will get annoyed by the lack of this kind of thing, anyway.

**Comment by marcphilipp** [comment]
Sorry, we haven't managed to discuss this PR in the team, yet. Hence, I reopened the PR.

**Comment by rmodini** [comment]
@marcphilipp thanks for reopening the pr! looking forward to this

**Comment by marcphilipp** [comment]
Did any of you have any concrete use cases in mind except for combining repeated and parameterized tests (when is that actually useful?). JUnit Pioneer has [`@CartesianProductTest`](https://junit-pioneer.org/docs/cartesian-product/). Does that address your use case?

**Comment by marcphilipp** [comment]
Also related: #871

**Comment by danielhodder** [comment]
The original use for this was for testing database migrations if I recall. We have versioned migrations (1.0, 1.2, 2.0, etc.) which are all cumulative. We also support multiple databases. So the idea was run tests after upgrading from each version to the latest, on al the supported databases.

You'd end up with tests like:

* Test the feature works on a freshly created Oracle database
* Test the feature work on a database that started on the 1.0 schema and was upgraded to the latest on MS SQL
and so on.

The main issue I see with things like the CartesianProductTest and other similar things, is they assume a simple value. In my case I was creating new databases for each test so there's a _lot_ of setup code to get that all going, and I think each execution of the template ends up registering about half a dozen specific ParameterResolvers and Before/AfterCallback hooks.

The basic goal was to have a test helper which let you focus on writing the test code, and not gluing together the test libraries by passing dynamic parameters off to the different test helpers yourself. Basically you go from:

```java
@EnumSource(DatabaseTypes.class)
@VersionDetector
public void test(DatabaseType database, String startingVersion) {
    try (Database d = helper.createDatabase(database)) {
        DataSource source = d.getDataSource()
        // Run some scripts
        scriptHelper.runScriptsFrom(startingVersion, d.getAdminDatasource());

        try (ApplicationPlatform a = platformHelper.initialize(source)) {
            // Now test
            assertTrue(a.foo());
        }
    }
}
```

To something more like

```
@ApplicationDatabaseTest
public void test(ApplicationPlatform a) {
    assertTrue(a.foo());
}
```

And the @ApplicationDatabaseTest annotation is a meta annotation which does all the setup, script running, and application initialization. Then provides the initialized platform with a ParameterResolver. It also deals with cleaning up the test databases, and the global state the ApplicationPlatform needs to work correctly.

Hope that helps give an example of where this kind feature is super helpful. It is possible to do today but you have to write the whole thing from scratch, rather than being able to assemble it out of the existing elements Junit provides.

**Comment by marcphilipp** [comment]
@danielhodder Thanks for the detailed description! I wonder if we should rather prioritize implementing #871 so that you could run the test class multiple times against different database types and the contained test methods against the different versions. Would that also work in your case?

**Comment by danielhodder** [comment]
Not so much for us. We don't want to have two tests that use the same database. Basically we want a brand new database set up for each test (isolation etc.). A lot of the applications need to setup things which would collide so it's easier to not.

In fairness though that's partly because that's how our test support has always worked, and so just porting the existing behavior is easier.

Another use case for use, which I thought just today is take the previous example, and then parameterize each test with a known set of Demo Data.

Basically #871 doesn't really change the overall landscape of only being able to template a test with one provider. The goal of this PR is to create a way of having smaller, re-usable, templating methods and then compose them as needed in your test; rather than having to create a custom extension for every test class.

Don't get me wrong though #871 would be awesome and that would definitely help for other use cases we have. (Bonus points if we can have a version of that at the Suite level so you can deploy servers with it.)

**Comment by ewoks** [comment]
Any updates here? Anything to be resolved before it lands to next release?

**Comment by danielhodder** [comment]
I've updated the copyright notices which were failing the spotless check.

**Comment by marcphilipp** [comment]
We've looked at this today in the team call. We're not yet convinced this would be easy to grasp from a user's perspective. Since #871 is less controversial and more generally useful, we think we should implement that first.

**Comment by danielhodder** [comment]
I don't think I'd argue that feature would be for very advanced usage only, and very much hard to grasp, if you didn't get it. I guess this exists to solve a specific problem which can't currently be handled in extensions. If there was a way of accomplishing this outside of the core engine I'd gladly write it there. As far as I can tell that's not possible though.

This would be a feature mostly of interest to large teams build big software packages which have a lot custom test extensions, which the team may not own. In our case most of these templates are managed by our 'platform' team, while the usages are in downstream app teams. If we can combine templates then teams don't have to choose between them.

I do thing #871 is a great step and might help if it's possible to use that with inner classes. That would let you have something like:

```java
class Test {
    @Parameterized
    class InnterTest { 
         @MyDatabaseTestTempate
         public void test(DataSource db) {
         }
    }
}
```

It's not as 'clean' but it accomplishes a similar end, although it does mean you have to have separate extensions for these, rather than re-using existing ones, which was my goal.

**Comment by danielhodder** [comment]
Just ran into a more simple example which might help demonstrate this one. I am trying to combine the `PamaterizedTest` annotation from JUnit 5, with the `RetryingTest` annotation from JUnit Pioneer. In the current implementation if I try and use the two together I won't get parameters on any of the runs which the retrying test template generates.

#871 would definitely help since it would also allow BeforeEach and AfterEach methods to be parameterized.

**Comment by danielhodder** [comment]
I spent a bit of time and wrote a basic version of this (with a lot of hacks) here: https://github.com/danielhodder/junit-test-template-combiner. If there's enough interest I'll figure out how to publish this to maven central.
